bokehjs/src/lib/models/glyphs/image.ts

```diff
@@ -48,6 +60,56 @@ export class ImageView extends XYGlyphView {
     }
   }
 
+  _index_data(): SpatialIndex {
+    const points = [];
+    for (let i = 0, end = this._x.length; i < end; i++) {
+      const [l, r, t, b] = this._lrtb(i);
+      if (isNaN(l+r+t+b) || !isFinite(l+r+t+b)) {
+        continue;
+      }
+      points.push({minX: l, minY: b, maxX: r, maxY: t, i});
+    }
+    return new SpatialIndex(points);
+  }
+
+  _lrtb(i: number) : [number, number, number, number]{
+    const l = this._x[i]
+    const r = l + this._dw[i]
+    const b = this._y[i]
+    const t = b + this._dh[i]
+    return [l, r, t, b]
+  }
+
+
+  _image_index(index : number, x: number, y : number) : ImageIndex {
+    const [l,r,t,b] = this._lrtb(index);
+    const width = this._width[index]
+    const height = this._height[index]
+    const dx = (r - l) / width;
+    const dy = (t - b) / height;
+    const dim1 = Math.floor((x-l) / dx)
+    const dim2 = Math.floor((y-b) / dy)
+    return {index:index, dim1:dim1, dim2:dim2, flat_index :((dim2 * width) + dim1)}
```

Comment:
Before this is merged, please clean up the new or modify code. I think I wouldn't be far off saying, that there are all combinations of whitespace involved with `:`, `,`, `-` and other symbols. Also, please remove trailing `;` and double empty lines.